[MSE] imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-negative.html is a permanent timeout

jyavenard opened the pull request:
#### a382af568e5ab5712408ef955f01b69af55364bd
<pre>
[MSE] imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-negative.html is a permanent timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=302693">https://bugs.webkit.org/show_bug.cgi?id=302693</a>
<a href="https://rdar.apple.com/164942613">rdar://164942613</a>

Reviewed by Youenn Fablet.

SourceBuffer.changeType was fundamentally broken on Apple platforms. The SourceBufferParser used
to demux the content was created and given to the SourceBufferPrivateAVFObjC
constructor. It could never be changed after that.
The code assumes that if we could create a SourceBufferParser for the required
content, then we supported the type given to changeType.
But that wasn&apos;t the case. If we had first created a SourceBufferParserAVFObjC
it could never demux webm content, same in the reverse case, but we reported
that it could.
The original scenario that lead to the changeType API was to be able to play
webm content and insert mp4 ads without having to recreate a new SourceBuffer.
That original scenario couldn&apos;t work.

We now allow to create a new SourceBufferParser whenever SourceBuffer.changeType
is called and replace the previously created one.

Re-enabled WPT tests disabled (since they were failing) since 2020.

* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-negative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-without-codecs-parameter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-util.js:
(trimBuffered): The test assumes that all videos used would be over 2.1s long, but they aren&apos;t they are 2.03s long only.
This caused an assertion to fire; Chrome failed in the same manner. The duration of the original file was irrelevant
to the test; all we need to do is not attempt to resize if the video is already shorter that request (2s)
(trimDuration): Lodged as <a href="https://github.com/web-platform-tests/wpt/issues/56347">https://github.com/web-platform-tests/wpt/issues/56347</a> ; fix submitted
(runChangeTypeTest):
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations: Remove failure/timeout expectations.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-expected.txt: Removed.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::createSourceBufferPrivate):
* Source/WebCore/platform/graphics/MediaSourceConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::addSourceBuffer):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::create): Save the original MediaSource configuration
used to enable TextTrack supports in MSE.
(WebCore::SourceBufferPrivateAVFObjC::SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::setTrackChangeCallbacks):
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal): We now need to ensure
we jump through the SourceBufferPrivate&apos;s queue first to ensure that events
run in the original queued order.
(WebCore::SourceBufferPrivateAVFObjC::resetParserStateInternal): Same as above.
(WebCore::SourceBufferPrivateAVFObjC::audioTrackDidChangeEnabled):
(WebCore::SourceBufferPrivateAVFObjC::canSwitchToType):
(WebCore::SourceBufferPrivateAVFObjC::configureParser): Added utility method.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/303665@main">https://commits.webkit.org/303665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80d8246e16badc32a783ac64536b07d7db7af789

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133178 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140731 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85220 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8b3db869-2923-4bc3-aaa7-d914089f5814) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135048 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6183 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101860 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69291 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2d4b85cd-377a-4180-9beb-7980bb4b9d25) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136125 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4382 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119362 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82656 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/01a71878-6b81-4a92-8995-6117fb6bd3af) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4267 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1843 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143380 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5349 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38057 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110238 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5431 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4599 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110420 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27988 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4140 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115619 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59076 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5404 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33973 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5250 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68856 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5493 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5360 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-util.js:
```diff
@@ -116,20 +116,23 @@ function appendBuffer(test, sourceBuffer, data) {
 }
 
-function trimBuffered(test, mediaElement, sourceBuffer, minimumPreviousDuration, newDuration, skip_duration_prechecks) {
+function trimBuffered(test, mediaElement, sourceBuffer, newDuration, skip_duration_prechecks) {
```
Please make a WPT corresponding PR if we need to upstream the changes (using export-w3c-tests script is the preferred way).

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
```diff
@@ -126,7 +127,7 @@ class SourceBufferPrivateAVFObjC final
 
 private:
-    explicit SourceBufferPrivateAVFObjC(MediaSourcePrivateAVFObjC&, Ref<SourceBufferParser>&&, Ref<AudioVideoRenderer>&&);
+    explicit SourceBufferPrivateAVFObjC(MediaSourcePrivateAVFObjC&, const MediaSourceConfiguration&, Ref<SourceBufferParser>&&, Ref<AudioVideoRenderer>&&);
```
```suggestion
   SourceBufferPrivateAVFObjC(MediaSourcePrivateAVFObjC&, const MediaSourceConfiguration&, Ref<SourceBufferParser>&&, Ref<AudioVideoRenderer>&&);
```

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -503,8 +473,11 @@
 
-    m_appendQueue->dispatch([parser = m_parser] {
-        parser->resetParserState();
+    ensureWeakOnDispatcher<SourceBufferPrivateAVFObjC>([](auto& buffer) {
```
```suggestion
    ensureWeakOnDispatcher([](auto& buffer) {
```
And remove the templated version which has no use AFAIk.

jyavenard replied:
It's definitely in use. It controls of what type the buffer parameter will be.

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -760,12 +733,64 @@
+        dispatcher->dispatch(WTFMove(function));
+    });
+
+    parser.setDidParseInitializationDataCallback([weakThis = ThreadSafeWeakPtr { *this }, dispatcher = m_dispatcher] (InitializationSegment&& segment) {
```
Do we need to capture m_dispatcher?
Could we instead reuse protectedThis->m_dispatcher instead?

jyavenard replied:
Reason this is done is that the parser runs in its own queue; it can be deleted after the SourceBufferPrivate. 
I guess we don't care about the results, but that also avoids doing two check on the ThreadSafeRefPtr (and the same one) for every callbacks.

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -96,50 +97,13 @@
 {
     ALWAYS_LOG(LOGIDENTIFIER);
 
+    configureParser(m_parser);
```
```suggestion
    setParserCallbacks(m_parser);
```

jyavenard replied:
I like configure better

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -96,50 +97,13 @@
 
+    configureParser(m_parser);
     if (isMainThread()) {
         if (RefPtr webmParser = dynamicDowncast<SourceBufferParserWebM>(m_parser)) {
```
`m_parser` can change when switching, should we redo that check every time we update it?

jyavenard replied:
That's really just to apply a quirk; it's non-standard, allowing sites to do the wrong thing under very particular circumstances. There's never going to use changeType or even MSE in a worker.

youennf reviewed (commented):
LGTM once there is a WPT PR and we validate whether we need to do the matroska check every time or a single time.

jyavenard committed:
[MSE] imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-negative.html is a permanent timeout
https://bugs.webkit.org/show_bug.cgi?id=302693
rdar://164942613

Reviewed by Youenn Fablet.

SourceBuffer.changeType was fundamentally broken on Apple platforms. The SourceBufferParser used
to demux the content was created and given to the SourceBufferPrivateAVFObjC
constructor. It could never be changed after that.
The code assumes that if we could create a SourceBufferParser for the required
content, then we supported the type given to changeType.
But that wasn't the case. If we had first created a SourceBufferParserAVFObjC
it could never demux webm content, same in the reverse case, but we reported
that it could.
The original scenario that lead to the changeType API was to be able to play
webm content and insert mp4 ads without having to recreate a new SourceBuffer.
That original scenario couldn't work.

We now allow to create a new SourceBufferParser whenever SourceBuffer.changeType
is called and replace the previously created one.

Re-enabled WPT tests disabled (since they were failing) since 2020.

* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-negative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-without-codecs-parameter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-changetype-util.js:
(trimBuffered): The test assumes that all videos used would be over 2.1s long, but they aren't they are 2.03s long only.
This caused an assertion to fire; Chrome failed in the same manner. The duration of the original file was irrelevant
to the test; all we need to do is not attempt to resize if the video is already shorter that request (2s)
(trimDuration): Lodged as https://github.com/web-platform-tests/wpt/issues/56347 ; fix submitted
(runChangeTypeTest):
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations: Remove failure/timeout expectations.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/mediasource-changetype-play-expected.txt: Removed.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::createSourceBufferPrivate):
* Source/WebCore/platform/graphics/MediaSourceConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::addSourceBuffer):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::create): Save the original MediaSource configuration
used to enable TextTrack supports in MSE.
(WebCore::SourceBufferPrivateAVFObjC::SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::setTrackChangeCallbacks):
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal): We now need to ensure
we jump through the SourceBufferPrivate's queue first to ensure that events
run in the original queued order.
(WebCore::SourceBufferPrivateAVFObjC::resetParserStateInternal): Same as above.
(WebCore::SourceBufferPrivateAVFObjC::audioTrackDidChangeEnabled):
(WebCore::SourceBufferPrivateAVFObjC::canSwitchToType):
(WebCore::SourceBufferPrivateAVFObjC::configureParser): Added utility method.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/303665@main
